src/config.ts

```diff
@@ -32,3 +52,50 @@ function getLocalDataPath() {
 
     return path.join(os.homedir(), ".mongodb", "mongodb-mcp");
 }
+
+// Gets the config supplied by the user as environment variables. The variable names
+// are prefixed with `MDB_MCP_` and the keys match the UserConfig keys, but are converted
+// to SNAKE_UPPER_CASE.
+function getEnvConfig(): Partial<UserConfig> {
+    const camelCaseToSNAKE_UPPER_CASE = (str: string): string => {
+        return str.replace(/([a-z])([A-Z])/g, "$1_$2").toUpperCase();
+    };
+
+    const result: Partial<UserConfig> = {};
+    Object.keys(defaults).forEach((key) => {
+        const envVarName = `MDB_MCP_${camelCaseToSNAKE_UPPER_CASE(key)}`;
+        if (process.env[envVarName]) {
+            result[key as keyof UserConfig] = process.env[envVarName];
+        }
+    });
+
+    return result;
+}
+
+// Gets the config supplied by the user as a JSON file. The file is expected to be located in the local data path
+// and named `config.json`.
+function getFileConfig(): Partial<UserConfig> {
+    const configPath = path.join(getLocalDataPath(), "config.json");
+
+    try {
+        const config = fs.readFileSync(configPath, "utf8");
+        return JSON.parse(config);
+    } catch {
+        return {};
+    }
+}
+
+// Merges several user-supplied configs into one. The precedence is from right to left where the last
+// config in the `partialConfigs` array overrides the previous ones. The `defaults` config is used as a base.
+function mergeConfigs(defaults: UserConfig, ...partialConfigs: Array<Partial<UserConfig>>): UserConfig {
+    const mergedConfig: UserConfig = { ...defaults };
+    for (const key of Object.keys(defaults)) {
+        for (const partialConfig of partialConfigs) {
+            if (partialConfig[key]) {
+                mergedConfig[key] = partialConfig[key];
+            }
+        }
+    }
+
+    return mergedConfig;
+}
```

Comment:
```suggestion
function mergeConfigs(defaults: UserConfig, ...partialConfigs: Array<Partial<UserConfig>>): UserConfig {
  return Object.assign({}, defaults, ...partialConfigs);
}
```

I believe this should do the same trick; we could get rid of the function also. 
The functional difference can be that if partial config had more field than the default those would also get assigned but considering they're meant to be partial, not a case we need to worry about.